File: internal/service/diagnostic.go
package service

import (
	"context"

	"uni-context/internal/port"
)

// DiagnosticService powers the `doctor` command. It owns the schema-version
// lookup (previously a raw `a.DB.QueryRow(...)` in the CLI) and the embedder
// health check (previously inline `a.Embedder.Embed` / `.Model()` calls).
// Routing these through a service means the inbound layer has no direct
// dependency on *sql.DB or port.Embedder, mirroring how ItemService owns
// the get/list/delete path.
type DiagnosticService struct {
	schema   port.SchemaMeta
	embedder port.Embedder // nil = Plan 1 (disabled); no PingEmbedder probe
}

// NewDiagnosticService wires the schema-meta reader and the optional
// embedder. embedder may be nil — PingEmbedder reports disabled in that
// case rather than attempting an Embed call.
func NewDiagnosticService(schema port.SchemaMeta, embedder port.Embedder) *DiagnosticService {
	return &DiagnosticService{schema: schema, embedder: embedder}
}

// SchemaVersion returns the migration version string from schema_meta.
// Errors propagate unwrapped so callers can distinguish a missing
// schema_meta table (uninitialized DB) from other failures.
func (s *DiagnosticService) SchemaVersion(ctx context.Context) (string, error) {
	return s.schema.Version(ctx)
}

// PingEmbedder exercises the live embedder with a one-token embed to
// surface transient failures (Ollama down, wrong base URL, auth reject)
// before they bite a real search. Returns:
//   - (zero, false, nil) when no embedder is wired (Plan 1). The CLI
//     uses enabled=false to print "disabled" rather than "FAIL".
//   - (ModelInfo, true, nil) when the embedder answered the ping. The
//     CLI prints "<slug>, <dim>-dim".
//   - (zero, true, err) when the embedder exists but the ping failed.
//     enabled stays true so the CLI prints "FAIL (...)", not "disabled".
//     Model() is intentionally NOT called on failure — matches the
//     previous inline doctor code and avoids masking the embed error
//     with a stale model label.
func (s *DiagnosticService) PingEmbedder(ctx context.Context) (port.ModelInfo, bool, error) {
	if s.embedder == nil {
		return port.ModelInfo{}, false, nil
	}
	if _, err := s.embedder.Embed(ctx, []string{"ping"}); err != nil {
		return port.ModelInfo{}, true, err
	}
	return s.embedder.Model(), true, nil
}

File: python/src/unictx/embed/diagnostic.py
"""DiagnosticService — powers the ``doctor`` command.

Behavior-port of Go's ``internal/service/diagnostic.go``. Owns the
schema-version lookup (previously raw ``a.DB.QueryRow(...)`` in the
CLI) and the embedder health check (previously inline
``a.Embedder.Embed`` / ``.model()`` calls). Routing these through a
service means the inbound layer has no direct dependency on the
storage connection or port.Embedder — mirroring how ItemService owns
the get/list/delete path.

Plan §Python Conventions adaptations:
  - ctx dropped (Python is sync).
  - Go's ``PingEmbedder(ctx) (ModelInfo, bool, error)`` →
    ``ping_embedder() -> tuple[ModelInfo, bool]``; on failure raises
    rather than returning the error as a third value. Callers wrap
    in try/except for the "FAIL" branch.
  - Embedder is optional (None = Plan 1 / disabled).

A consumer-side ``SchemaMeta`` Protocol is defined locally — the
Phase 2 ``SchemaMetaImpl`` is the only known implementation, but
introducing the Protocol keeps the dependency structural (any class
with a ``version() -> str`` method satisfies it) and lets tests
inject stubs without subclassing the concrete impl.
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from unictx.embed.embedder import Embedder, ModelInfo

__all__ = ["DiagnosticService", "SchemaMeta"]


@runtime_checkable
class SchemaMeta(Protocol):
    """Read-only accessor for ``schema_meta.schema_version``.

    Consumer-side Protocol — the concrete :class:`SchemaMetaImpl` is
    the production impl. Defined here (vs in storage/) because the
    consumer (DiagnosticService) owns what it needs; this matches the
    IngestService._Embedder pattern.
    """

    def version(self) -> str:
        """Return the migration version string.

        Raises SchemaMetaNotFound if the ``schema_version`` row is
        missing (uninitialized DB). Other failures propagate verbatim.
        """
        ...


class DiagnosticService:
    """Powers the ``doctor`` command.

    Construction is cheap (no I/O). ``embedder`` is optional — when
    None, ``ping_embedder`` reports disabled rather than attempting
    an embed call. This mirrors Plan 1 behavior: no embedder wired →
    ``doctor`` says "disabled" instead of "FAIL".
    """

    def __init__(
        self,
        schema: SchemaMeta,
        embedder: Embedder | None = None,
    ) -> None:
        self._schema = schema
        self._embedder = embedder

    def schema_version(self) -> str:
        """Return the migration version string.

        Errors propagate unwrapped so callers can distinguish a missing
        schema_meta table (uninitialized DB) from other failures.
        """
        return self._schema.version()

    def ping_embedder(self) -> tuple[ModelInfo, bool]:
        """Exercise the embedder with a one-token embed.

        Returns:
          - ``(zero, False)`` when no embedder is wired (Plan 1).
            Callers use ``enabled=False`` to print "disabled" rather
            than "FAIL".
          - ``(ModelInfo, True)`` when the embedder answered the ping.
            Callers print "<slug>, <dim>-dim".
          - On failure: raises the embedder's exception. ``enabled``
            stays True (caller catches the exception to print
            "FAIL (...)").

        ``model()`` is intentionally NOT called on failure — matches
        Go's behavior and avoids masking the embed error with a stale
        model label.
        """
        if self._embedder is None:
            return ModelInfo(), False
        # One-token embed to surface transient failures (Ollama down,
        # wrong base URL, auth reject) before they bite a real search.
        self._embedder.embed(["ping"])
        return self._embedder.model(), True
